core/src/main/java/io/questdb/cairo/mv/SampleByIntervalIterator.java

```diff
@@ -24,52 +24,108 @@
 
 package io.questdb.cairo.mv;
 
+import io.questdb.griffin.engine.groupby.TimestampSampler;
+import io.questdb.std.LongList;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
 /**
- * Allows iterating through SAMPLE BY buckets with the given step.
- * The goal is to split a potentially large time interval to be scanned by
- * the materialized view query into smaller intervals, thus minimizing
- * chances of out-of-memory kills.
+ * Iterates through SAMPLE BY buckets with the given step. The goal is to split
+ * a potentially large time interval to be scanned by the materialized view query
+ * into smaller intervals, thus minimizing chances of out-of-memory kills.
+ * <p>
+ * When the list of txn min-max timestamp intervals is given, the iterated
+ * SAMPLE BY buckets that have no intersection with the intervals,
+ * i.e. remain unchanged, will be skipped.
  */
-public interface SampleByIntervalIterator {
+public abstract class SampleByIntervalIterator {
+    protected TimestampSampler sampler;
+    protected int step;
+    // index of next txn timestamp interval to check against
+    private int txnIntervalLoIndex;
+    private LongList txnIntervals;
 
     /**
      * Returns maximum timestamp that belong to the iterated intervals.
      */
-    long getMaxTimestamp();
+    public abstract long getMaxTimestamp();
 
     /**
      * Returns minimum timestamp that belong to the iterated intervals.
      */
-    long getMinTimestamp();
+    public abstract long getMinTimestamp();
 
     /**
      * Returns minimal number of SAMPLE BY buckets for in a single iteration.
      */
-    int getStep();
+    public abstract int getStep();
 
     /**
      * High boundary for the current iteration's interval.
      * Meant to be used as an exclusive boundary when querying.
      */
-    long getTimestampHi();
+    public abstract long getTimestampHi();
 
     /**
      * Low boundary for the current iteration's interval.
      * Meant to be used as an inclusive boundary when querying.
      */
-    long getTimestampLo();
+    public abstract long getTimestampLo();
 
     /**
      * Iterates to the next interval.
      *
      * @return true if the iterator moved to the next interval; false if the iteration has ended
      */
-    boolean next();
+    public boolean next() {
+        OUT:
+        while (next0()) {
+            if (txnIntervals != null) {
+                for (int n = txnIntervals.size(); txnIntervalLoIndex < n; txnIntervalLoIndex += 2) {
+                    final long intervalLo = txnIntervals.getQuick(txnIntervalLoIndex);
+                    final long intervalHi = txnIntervals.getQuick(txnIntervalLoIndex + 1);
+                    final long iteratorLo = getTimestampLo();
```

Comment:
nit: `iteratorLo`, `iteratorHi` do not change inside the for loop, do they? You can move the outside of the for loop.

Reply:
Done in 5a3dfd12ebb85c92f7476538dcf7e0ce8688aac6